### PR TITLE
Experiment BT support using Chrome web serial

### PR DESCRIPTION
**Pull request by pelikhan**
Use web serial (https://wicg.github.io/serial/) in latest Chrome/Edge Insider with Experimental Web Features enabled (chrome://flags/#enable-experimental-web-platform-features or edge://flags/#enable-experimental-web-platform-features ) to transfer the program over Bluetooth.

* diff no spaces: https://github.com/microsoft/pxt-ev3/pull/920/files?w=1
* test build: https://makecode.mindstorms.com/app/8311d0f5bc212ac6c0482ed621c7228c67df0181-a7a4ff97e0

image

- [x] handle wrong com port
- [x] detect serial support
- [x] add button in download dialog
- [x] disable BT if download fails

### CODE REVIEW

File: editor/deploy.ts [hunk 1]
```diff
@@ -2,57 +2,190 @@
...
+    return Promise.resolve();
+}
+
+export function canUseWebSerial() {
+    return WebSerialPackageIO.isSupported();
+}
+
+export function enableWebSerial() {
+    initPromise = undefined;
+    useHID = useWebSerial = WebSerialPackageIO.isSupported();
```
Comment (riknoll) [inline editor/deploy.ts]: Nitpick: split these lines

File: editor/deploy.ts [hunk 2]
```diff
@@ -2,57 +2,190 @@
...
+function initHidAsync() { // needs to run within a click handler
+    if (initPromise)
+        return initPromise
+    if (useHID) {
         initPromise = hf2Async()
             .catch(err => {
+                console.error(err);
                 initPromise = null
-                noHID = true
+                useHID = true
```
Comment (riknoll) [inline editor/deploy.ts]: Should this be inverted?

File: editor/wrap.ts [hunk 3]
```diff
@@ -1,285 +1,281 @@
...
+    talkAsync(buf: Uint8Array, altResponse = 0) {
+        return this.lock.enqueue("talk", () => {
+            this.msgs.drain()
+            if (this.dataDump)
+                log("TALK: " + U.toHex(buf))
+            return this.io.sendPacketAsync(buf)
+                .then(() => this.msgs.shiftAsync(1000))
+                .then(resp => {
+                    if (resp[2] != buf[2] || resp[3] != buf[3])
+                        U.userError("msg count de-sync")
```
Comment (riknoll) [inline editor/wrap.ts]: Should these be surfaced to the user?

File: editor/extension.ts [hunk 4]
```diff
@@ -84,7 +74,14 @@ pxt.editor.initExtensionsAsync = function (opts: pxt.editor.ExtensionOptions): P
                 hideAgree: false,
                 agreeLbl: lf("I got it"),
                 className: 'downloaddialog',
-                buttons: [downloadAgain ? {
+                buttons: [canUseWebSerial ? {
```
Comment (riknoll) [inline editor/extension.ts]: I think `canUseWebSerial` is a function.

File: editor/wrap.ts [hunk 5]
```diff
@@ -1,285 +1,281 @@
...
+                            type: 'serial',
+                            id: 'n/a', // TODO?
+                            data: str
+                        }, "*")
+                } else
+                    console.log("Magic: " + code + ": " + U.toHex(payload))
+                return
+            }
+            if (this.dataDump)
+                log("RECV: " + U.toHex(buf))
```
Comment (riknoll) [inline editor/wrap.ts]: Difference between `log()` and `console.log()` (used above)?

File: editor/wrap.ts (hunk 3 above)
Comment (pelikhan) [reply editor/wrap.ts]: they get swallowed later on

File: editor/wrap.ts (hunk 5 above)
Comment (pelikhan) [reply editor/wrap.ts]: just a log helper for this file. this is mostly the same code as before, but reindented.

Comment (pelikhan) [reply editor/wrap.ts]: reduce log level to debug

File: docs/bluetooth.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+## ~ hint
+
+### WARNING: EXPERIMENTAL FEATURES AHEAD! 
+
+Support for Bluetooth download relies on [Web Serial](https://wicg.github.io/serial/),
+an experimental browser feature. Web Serial is a work [in progress](https://www.chromestatus.com/feature/6577673212002304); 
+it may change or be removed in future versions without notice.
+
+By enabling these experimental browser features, you could lose browser data or compromise your security
```
Comment (mmoskal) [inline docs/bluetooth.md]: your -> security of devices connected to your computer?

Comment (pelikhan) [reply docs/bluetooth.md]: it's the verbiage of chrome
image

**Comment by ngbrown** [comment]
Any chance of this working with the Electron based offline version?  It worked in Chrome, but the newly released 1.2.26 offline app doesn't give the Bluetooth download option.

**Pull request merged**